internal/discovery/service.go

```diff
@@ -0,0 +1,142 @@
+}
+
+// Create lookup key based on input request
+// Using the sorted nature of keys, lookup by either database or by database and vBucket or all keys
+func createKeyToGet(request *serverpb.GetClusterInfoRequest) []byte {
+	if request.GetDatabase() != "" {
+		if request.GetVBucket() != "" {
+			return []byte(request.GetDatabase() + ":" + request.GetVBucket() + ":")
+		} else {
+			return []byte(request.GetDatabase() + ":")
+		}
+	} else {
+		return []byte{}
+	}
+}
+
+func (d *discoverService) UpdateStatus(ctx context.Context, request *serverpb.UpdateStatusRequest) (*serverpb.Status, error) {
+	// Store the serialised form of the request
+	val, err := json.Marshal(request)
+
+	if err != nil {
+		d.logger.Error("Unable to marshal status request", zap.Error(err))
+		return newErrorStatus(err), err
+	}
+
+	// TODO - check if expiry TTL is working
+	putRequest := serverpb.PutRequest{Key: createKeyToInsert(request.GetRegionInfo()), Value: val, ExpireTS: hlc.GetUnixTimeFromNow(d.config.StatusTTl)}
+	if _, err = d.dkvCli.Put(ctx, &putRequest); err != nil {
+		return newErrorStatus(err), err
+	} else {
+		return &serverpb.Status{Code: 0, Message: ""}, nil
+	}
+}
+
+func (d *discoverService) GetClusterInfo(ctx context.Context, request *serverpb.GetClusterInfoRequest) (*serverpb.GetClusterInfoResponse, error) {
+	iterReq := &serverpb.IterateRequest{KeyPrefix: createKeyToGet(request)}
+	kvStrm, err := d.dkvCli.Iterate(ctx, iterReq)
+	if err != nil {
+		d.logger.Error("Unable to get cluster info", zap.Error(err))
+		return nil, err
+	}
+	var clusterInfo []storage.KVEntry
+	for {
+		itRes, err := kvStrm.Recv()
+		if err == io.EOF {
+			break
+		} else if (err != nil) {
+			// Better to return error rather than partial results as incomplete results could cause client misbehaviour
+			d.logger.Error("Partial failure in getting cluster info", zap.Error(err))
+			return nil, err
+		} else {
+			clusterInfo = append(clusterInfo, storage.KVEntry{
+				Key:   itRes.Key,
+				Value: itRes.Value,
+			})
+		}
+	}
+
+	var regionsInfo []*serverpb.RegionInfo
+	for _, serializedStatusUpdate := range clusterInfo {
+		statusUpdate := serverpb.UpdateStatusRequest{}
+		if err = json.Unmarshal(serializedStatusUpdate.Value, &statusUpdate); err != nil {
+			d.logger.Error("Unable to unmarshal status request", zap.Error(err))
+			continue
+		}
+		// Filter regions outside the requested DC (if provided)
+		if request.GetDcID() != "" && request.GetDcID() != statusUpdate.GetRegionInfo().GetDcID() {
+			continue
+		}
+		// Filter inactive regions and regions whose status was updated long time back and hence considered inactive
+		if hlc.GetTimeAgo(statusUpdate.GetTimestamp()) < d.config.HeartbeatTimeout && statusUpdate.GetRegionInfo().GetStatus() != serverpb.RegionStatus_INACTIVE {
+			regionsInfo = append(regionsInfo, statusUpdate.GetRegionInfo())
+		}
+		// TODO : Filter such that only 1 master is returned per region
```

Comment:
@krishan1390 Check this TODO

Reply:
it is a TODO but not important right now as we don't depend on if a node is master / follower. this requires first fixing the in memory logic of listNodes() and then sorting by timestamp in this code to return just 1 master.

Reply:
Please check the updated response of ListNodes

Reply:
have added a comment in https://github.com/flipkart-incubator/nexus/commit/32ce0a3d33917cf08177f13d4d594849e31499c0 

Let me know if im mistaken here

Reply:
Actually not. The leadership information is always up to date. Please check my detailed response on your comment